jena-arq/src/main/java/org/apache/jena/sparql/core/TxnDataset2Graph.java

```diff
@@ -0,0 +1,240 @@
+        Object key = calcKey(graph);
+        handlers.remove(graph);
+    }
+
+    // Attempt to manage the graph transactions during a transaction.
+    // Imperfect for removal, we don't know whether to call commit() or abort().
+    // Works for adding.
+    // Generally better not to change the graphs during a transaction, just set them once
+    // on creation.
+    
+    public void addGraph(Graph graph) {
+        checkNotReadMode();
+        if ( graph == null )
+            return;
+        if ( ! handlers.containsKey(graph) ) {
+            // Add if new.
+            addHandler(handlers, graph) ;
+            if ( super.isInTransaction() ) {
+                // If we are in a transaction, start the subtransaction. 
+                TransactionHandler th = handlers.get(graph);
+                if ( th != null )
+                    th.begin();
+            }
+        }
+    }
+    
+    public void removeGraph(Graph graph) {
+        checkNotReadMode();
+        if ( graph == null )
+            return;
+        if ( ! super.isInTransaction() ) {
+            // Not in transaction, do now. 
+            removeHandler(handlers, graph);
+            return;
+        }
+        // Queue to be removed at the end.
+        Set<Graph> toBeRemoved = removedGraphs.get();
+        if ( toBeRemoved == null ) {
+            // Lazy set of the HashSet. 
+            toBeRemoved = new HashSet<>();
+            removedGraphs.set(toBeRemoved);
+        }
+        removedGraphs.get().add(graph);
+    }
+
+    public void setPrimaryGraph(Graph graph) {
+        checkNotReadMode();
+        if ( graph == null )
+            return;
+        removeGraph(graph);
+        addGraph(graph);
+    }
+    
+    private void handlers(Consumer<TransactionHandler> action) {
+        synchronized (lock) {
+            handlers.forEach((g,th)->action.accept(th));
+        }
+    }
+
+    private void checkNotReadMode() {
+        if ( !super.isInTransaction() )
+            return;
+        if ( super.isTransactionMode(ReadWrite.READ) )
+            throw new JenaTransactionException("In READ mode in transaction");
+    }
+
+    private ThreadLocal<Set<Graph>> removedGraphs = ThreadLocal.withInitial(()->null);
+    private void start() {}
+    private void finish() {
+        if ( ! super.isTransactionMode(ReadWrite.WRITE) )
+            return;
+        // This is called inside the lock of super.
+        Set<Graph> toBeRemoved = removedGraphs.get();
+        removedGraphs.remove();
+        if ( toBeRemoved == null )
+            return ;
+        toBeRemoved.forEach(g->removeHandler(handlers, g));
+    }
+    
+    // TransactionalLock.begin(ReadWrite) calls begin(TxnType)
+    @Override
+    public void begin(TxnType type) {
+        super.begin(type);
+        // Whatever the type. Graph Transactions do not allow for "read-only".
+        start();
+        handlers(h->h.begin());
```

Comment:
No big deal, but this and things like it could be `handlers(TransactionHandler::begin)`, which I think reads a bit easier.

Reply:
I don't think it always does :-) `h.` means the IDE method finding works.

Reply:
It's fine not to, but what kind of IDE are you using that can't trace method references? Eclipse certainly can.

Reply:
Finding for completion on the lambda argument.  Complete on `handlers(` is `Consumer<TransactionHandler> action`.

Reply:
Yeah, Eclipse (Cmd-space) gives me that, no problem.